### DISCUSSION TITLE
handling of empty lists langium 3.0 vs langium 2.0

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#1535] [Q&A] [answered]
I have observed an effect (not critical) and would like to clarify if this is some bug hidden in my code somewhere or an effect to be expected:

I have a grammar as follows, defining a list of strings - or nothing (`(...)?`):
```
('notes' '=' notes+=STRING (',' notes+=STRING)*)?
```

* In langium 2.x I got always an array `notes`, which had a length of 0 in some cases
* In langium 3.x I get `undefined` if no notes are present (and not an array of length 0).

Is my observation in langium 3.x (`undefined` if no notes are present) the expected behavior?

**Comment by msujew** [comment]
Hey @goto40,

This actually sounds like a bug/regression. It's still supposed to be an empty list. Can you share a reproducible example so I can check what's wrong?

**Reply by goto40** [reply]
Jupp. I will try to create a small example.

**Reply by goto40** [reply]
When creating a hello-world language (`yo langium`) you can obtain the effect with:

The following grammar:
```
grammar HelloWorld

entry Model:
    (persons+=Person | greetings+=Greeting)*;

Person:
    'person' name=ID ('notes' '=' notes+=STRING (',' notes+=STRING)*)?;  // <----- modified

Greeting:
    'Hello' person=[Person:ID]'!';

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;
terminal INT returns number: /[0-9]+/;
terminal STRING: /"(\\.|[^"\\])*"|'(\\.|[^'\\])*'/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

and a generator 
```
    const fileNode = expandToNode`
        "use strict";

        ${model.persons.map(p=>`${p.name}: ${p.notes.join(',')}`).join('\n')}  // <-- modified
        ${joinToNode(model.greetings, greeting => `console.log('Hello, ${greeting.person.ref?.name}!');`, { appendNewLineIfNotEmpty: true })}
    `.appendNewLineIfNotEmpty();
```

The following model works as expected:
```
person Pierre notes = "note1", "note2"
person Tim notes = "n1"
person Tom notes = "note"
```

But w/o notes for the second person you get an error while generating code for it
```
person Pierre notes = "note1", "note2"
person Tim // No notes (notes-list should ne empty, not undefined)
person Tom notes = "note"
```

Result of running the generator:
```
node .\bin\cli.js generate .\examples\hello.hello
file:///C:/Disk_J/MDSD/experiments/hello-world/out/cli/generator.js:11
        ${model.persons.map(p => `${p.name}: ${p.notes.join(',')}`).join('\n')}
                                                       ^

TypeError: Cannot read properties of undefined (reading 'join')
    at file:///C:/Disk_J/MDSD/experiments/hello-world/out/cli/generator.js:11:56
    at Array.map (<anonymous>)
    at generateJavaScript (file:///C:/Disk_J/MDSD/experiments/hello-world/out/cli/generator.js:11:25)
    at Command.generateAction (file:///C:/Disk_J/MDSD/experiments/hello-world/out/cli/main.js:17:31)

Node.js v20.10.0
```

**Reply by msujew** [reply, marked as answer]
Thanks for the repro example! I can reproduce this on the playground and on langium 3.0 itself, but apparently I fixed it as a driveby of https://github.com/eclipse-langium/langium/pull/1473. It's no longer reproducible on `main`. The fix will be released with 3.1 soonish.

**Reply by goto40** [reply]
Thank you for your work!
